addons/account/models/account_reconcile_model.py

```diff
@@ -355,267 +367,372 @@ def _prepare_reconciliation(self, st_line, aml_ids=[], partner=None):
+                            break
+
+        return results
+
+    @api.model
+    def _get_models_to_flush_for_apply_rules(self):
+        """ Returns a list of all the model names that need to be flushed
+        before matching the reconiliation models (so, the ones that will be called
+        from SQL queries).
+        """
+        return [
+            'account.bank.statement',
+            'account.bank.statement.line',
+            'account.move',
+            'account.move.line',
+            'res.company',
+            'account.journal',
+            'account.account',
+        ]
+
+    def _is_applicable_for(self, st_line, partner):
+        """ Returns true iff this reconciliation model can be used to search for matches
+        for the provided statement line and partner.
+        """
         self.ensure_one()
-        rule = self
-        # Filter on journals.
-        if rule.match_journal_ids:
-            query += ' AND st_line_move.journal_id IN %s'
-            params += [tuple(rule.match_journal_ids.ids)]
-
-        # Filter on amount nature.
-        if rule.match_nature == 'amount_received':
-            query += ' AND st_line.amount >= 0.0'
-        elif rule.match_nature == 'amount_paid':
-            query += ' AND st_line.amount <= 0.0'
-
-        # Filter on amount.
-        if rule.match_amount:
-            query += ' AND ROUND(ABS(st_line.amount), jnl_precision.dp) '
-            if rule.match_amount == 'lower':
-                query += '< %s'
-                params += [self.match_amount_max]
-            elif rule.match_amount == 'greater':
-                query += '> %s'
-                params += [self.match_amount_min]
-            else:
-                # if self.match_amount == 'between'
-                query += 'BETWEEN %s AND %s'
-                params += [rule.match_amount_min, rule.match_amount_max]
+
+        # Partner mapping rules are only applicable if
+        if self.rule_type == 'partner_mapping':
+            return not partner
+
+        # Filter on journals, amount nature, amount and partners
+        if ((self.match_journal_ids and st_line.move_id.journal_id not in self.match_journal_ids)
+            or (self.match_nature == 'amount_received' and st_line.amount < 0)
+            or (self.match_nature == 'amount_paid' and st_line.amount > 0)
+            or (self.match_amount == 'lower' and abs(st_line.amount) >= self.match_amount_max)
+            or (self.match_amount == 'greater' and abs(st_line.amount) <= self.match_amount_min)
+            or (self.match_amount == 'between' and (abs(st_line.amount) > self.match_amount_max or abs(st_line.amount) < self.match_amount_min))
+            or (self.match_partner and not partner)
+            or (self.match_partner and self.match_partner_ids and partner not in self.match_partner_ids)
+            or (self.match_partner and self.match_partner_category_ids and not partner.category_id & self.match_partner_category_ids)
```

Comment:
`AND partner.category_id AND partner.category_id not in self.match_partner_category_ids`?

Reply:
AND partner.category_id AND partner.category_id not in self.match_partner_category_ids
Will return False in case there isn't any category set on the partner, so it means the model will be applicable for them even if it restricted its application to a set of categories. It's wrong.

Though, reading it again, I think we actually could simply do 
partner.category_id not in self.match_partner_category_ids
in this case.